### ISSUE TITLE
gogs in Docker container uses Host IP trying to access linked DB container

### ISSUE THREAD

**Issue by nem75** [gogits/gogs#2015]
Hi, I'm trying to run gogs with this `docker-compose.yml`, linking it with a MariaDB container:

    gogs:
      image: gogs/gogs
      ports:
        - "10022:22"
        - "8080:3000"
      links:
        - db

    db:
      image: mariadb
      environment:
        - MYSQL_ROOT_PASSWORD=rootpass
        - MYSQL_DATABASE=gogs
        - MYSQL_USER=gogs
        - MYSQL_PASSWORD=gogs

On `localhost:8080` I configure gogs to use these database settings, but always get the error message shown:

![screenshot from 2015-11-22 11-47-02](https://cloud.githubusercontent.com/assets/1327785/11323273/e60bd908-910e-11e5-8b9e-2d4ed87716f8.png)

`77.37.17.149` is the IP of my router, which means that inside the gogs container a hostname is used for the DB which cannot be resolved (which then, via a catch-all, is resolved to the router). This is supported by the fact that the linked `db` container resolves (correctly) to a `172.*` address from inside the gogs container:

    $ docker exec -ti gogs_gogs_1 bash
    bash-4.3# ping db
    PING db (172.17.0.2): 56 data bytes
    64 bytes from 172.17.0.2: seq=0 ttl=64 time=0.404 ms
    64 bytes from 172.17.0.2: seq=1 ttl=64 time=0.185 ms
    ^C

Also, MariaDB is clearly reachable from within the gogs container:

    bash-4.3# telnet db 3306
    5.5.5-10.0.21-MariaDB-1~jessie1`(....)

So what hostname does the gogs container use trying to connect to the database? It's not `db`, or else it would work. Any ideas?

**Comment by Unknwon** [comment]
Hi,

What if you directly use `172.17.0.2`?

**Comment by Unknwon** [comment]
cc @0rax

**Comment by nem75** [comment]
Yeah, using the IP works. (Not really a sustainable solution, as I'm sure you know.)

Any idea what's wrong with using the hostname? `docker-compose logs` didn't give me no hint either ...

**Comment by Unknwon** [comment]
Maybe the way Gogs gets started does not receive the correct value of `db`?

**Comment by nem75** [comment]
The `links` section in the `docker-compose` links the containers the same as the `--link` option of `docker run`.

In this case it means that by entries in the gogs container's `/etc/host` and various environment variables, `db` inside the gogs container resolves to the correct IP.

As shown in my original post (see the `ping` and `telnet` output above).

So IMO gogs does not use the `db:3306` entered into the configuration field correctly, but I cannot for the life of me figure out why or what it does wrong there.

**Comment by 0rax** [comment]
@nem75 sorry for the lateness of my comment. 

From your test it seems that the issue comes from gigs itself and not the docker container itself? 

A solution to your problem would be to use localhost:db-port to access your Database. Linked container are bind to localhost using socat as this makes configuration cleaner imho.

**Comment by Unknwon** [comment]
Hmm... Gogs passes db info to driver: https://github.com/gogits/gogs/blob/master/models/models.go#L125-L132 Didn't take a look how the driver resolves the hosts.

**Comment by 0rax** [comment]
@nem75 in this case you may try to use `127.0.0.1` instead of localhost or db as db should be linked to localhost using socat. You should have as the first line of `gogs` docker container logs something like.

```
init:socat  | Linked container POSTGRES will be binded to localhost on port 5432
```

**Comment by Unknwon** [comment]
ANy updates from author of the issue?

**Comment by Unknwon** [comment]
Close now but open for continuing discussion.

**Comment by arantes555** [comment]
Hi, having the same issue here, for both postgres and ldap...
Worked around it by using the 127.0.0.1 method (thanks 0rax), but it does seem like a hack..

**Comment by nem75** [comment]
Ok, I understand that it's an issue because of socat and that there are workarounds. Correct?

It's just that from a Docker point of view those workarounds should not be necessary - if you link two containers or put them in the same docker network, they should be able to resolve routes to each other via their "docker hostnames" no matter what. Requiring the user to use IPs or to bind the container port to the host system is not the Docker way as it makes using the containers much less flexible.

Just saying. Of course it's up to you to decide if you think this relevant. :)

Out of curiosity, what exactly do you do with socat (haven't used it so far) and why?

**Comment by 0rax** [comment]
Hi @nem75,

I totally agree with you that this workaround should not be necessary (and it is here as an alternative / you can deactivate it at any time) and that container should be able to resolve route to others.

On the other hand `socat` is just used as a relay and bind an external socket to a local one making just a bridge between them. 

After all I dont really know why the Gogs installer is not able to resolve hosts at all while the underlying Docker container works as expected (just re-tested and I can access my database using the hostname present in `/etc/hosts`).

**Comment by Unknwon** [comment]
@0rax maybe the way [MySQL driver in Go](https://github.com/go-sql-driver/mysql) has some limitations?

**Comment by Unknwon** [comment]
@0rax maybe the way [MySQL driver in Go](https://github.com/go-sql-driver/mysql) has some limitations?

**Comment by 0rax** [comment]
Maybe but it seems to happen also with Postgre so I dont really know from where this come. I will try to do further testing tomorrow on a brand new CoreOS VM. 

Will keep you informed here.

**Comment by Unknwon** [comment]
@0rax probably it is caused by Go's bulitin host lookup?

**Comment by arantes555** [comment]
I have the problem, like I said, with both openLdap and Postegres, so I do
believe it cannot be a MySQL problem. Might be a problem caused by Go's
built-in host lookup, yeah. And if that is indeed the case, the Gogs devs
should definitely bump the report up to the Go devs.

2015-12-08 13:47 GMT-08:00 无闻 <notifications@github.com>:

**Comment by cweagans** [comment]
I'm having this problem too. Using the docker hostname in a PHP app definitely works, so it's not a Docker problem. Gogs is the only application I've ever had this problem with (but it's also the only Go app that I've ever deployed with Docker too).

**Comment by cweagans** [comment]
Related: https://github.com/drone/drone/issues/1269

**Comment by 0rax** [comment]
Hi there, sorry for my inactivity this past month.

@cweagans I've tested the solution provided in your link and it seems to work perfectly. By that I mean I was able to make Gogs resolve pgsql to the ip of my database removing the need for the socat link.

I will push a fix to provide name lookup to the go driver making evrybody able to reference their databases as their dns name instead of needing an socat link !

**Comment by Unknwon** [comment]
This is claimed to be fixed by merging #2475

**Comment by Unknwon** [comment]
Close as fixed.
